src/read_input_stream.rs

```diff
@@ -0,0 +1,307 @@
+                None => {
+                    return Err(glib::Error::new(
+                        crate::IOErrorEnum::Closed,
+                        "Already closed",
+                    ));
+                }
+                Some(Reader::Read(ref mut read)) => read,
+                Some(Reader::ReadSeek(ref mut read)) => read,
+            };
+
+            loop {
+                match std_error_to_gio_error(read.read(buffer)) {
+                    None => continue,
+                    Some(res) => return res,
+                }
+            }
+        }
+
+        fn close(
+            &self,
+            _stream: &InputStream,
+            _cancellable: Option<&crate::Cancellable>,
+        ) -> Result<(), glib::Error> {
+            let _ = self.read.borrow_mut().take();
+            Ok(())
+        }
+    }
+
+    impl SeekableImpl for ReadInputStream {
+        fn tell(&self, _seekable: &crate::Seekable) -> i64 {
+            // Not stream_position is not stable yet
+            // let mut read = self.read.borrow_mut();
+            // match *read {
+            //     Some(Reader::ReadSeek(ref mut read)) => {
+            //         read.stream_position().map(|pos| pos as i64).unwrap_or(-1)
+            //     },
+            //     _ => -1,
+            // };
+            -1
+        }
+
+        fn can_seek(&self, _seekable: &crate::Seekable) -> bool {
+            let read = self.read.borrow();
+            match *read {
+                Some(Reader::ReadSeek(_)) => true,
+                _ => false,
+            }
+        }
+
+        fn seek(
+            &self,
+            _seekable: &crate::Seekable,
+            offset: i64,
+            type_: glib::SeekType,
+            _cancellable: Option<&crate::Cancellable>,
+        ) -> Result<(), glib::Error> {
+            use std::io::SeekFrom;
+
+            let mut read = self.read.borrow_mut();
+            match *read {
+                Some(Reader::ReadSeek(ref mut read)) => {
+                    let pos = match type_ {
+                        glib::SeekType::Cur => SeekFrom::Current(offset),
+                        glib::SeekType::Set => {
+                            if offset < 0 {
+                                return Err(glib::Error::new(
+                                    crate::IOErrorEnum::InvalidArgument,
+                                    "Invalid Argument",
+                                ));
+                            } else {
+                                SeekFrom::Start(offset as u64)
+                            }
+                        }
+                        glib::SeekType::End => SeekFrom::End(offset),
+                        _ => unimplemented!(),
+                    };
+
+                    loop {
+                        match std_error_to_gio_error(read.seek(pos)) {
+                            None => continue,
+                            Some(res) => return res.map(|_| ()),
```

Comment:
Here we need to also check the `Ok` case: the actual position seeked to might be different than the requested position.

Reply:
Actually no: if it succeeds then it is at the requested position. The return position is only there to get an absolute value for the new position in case a seek relative to the end or current position was done.